tor/controller.go

```diff
@@ -168,26 +189,73 @@ func parseTorReply(reply string) map[string]string {
 }
 
 // authenticate authenticates the connection between the controller and the
-// Tor server using the SAFECOOKIE or NULL authentication method.
+// Tor server using the SAFECOOKIE authentication method.
 func (c *Controller) authenticate() error {
+	protocolInfo, err := c.protocolInfo()
+	if err != nil {
+		return err
+	}
+
+	// With the version retrieved, we'll cache it now in case it needs to be
+	// used later on.
+	c.version = protocolInfo.version()
+
+	switch {
+	// If a password was provided, then we should attempt to use the
+	// HASHEDPASSWORD authentication method.
+	case c.password != "":
+		if protocolInfo.supportsAuthMethod(authHashedPassword) {
+			return c.authenticateViaHashedPassword()
+		}
+
+		return fmt.Errorf("%v authentication method not supported",
```

Comment:
Probably better to flip this, to have ["the happy path"](https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88) aligned more to left.